File: sdk/ruby/test/test_client.rb
require "minitest/autorun"
require "json"
require "webrick"
require_relative "../lib/gradience"

class TestClient < Minitest::Test
  def setup
    @server = WEBrick::HTTPServer.new(Port: 0, Logger: WEBrick::Log.new("/dev/null"), AccessLog: [])
    @server.mount_proc "/api/wallets" do |req, res|
      res.content_type = "application/json"
      if req.request_method == "POST"
        res.body = { id: "w1", name: "demo", status: "active", created_at: "2026-01-01T00:00:00Z" }.to_json
      else
        res.body = [{ id: "w1", name: "demo", status: "active", created_at: "2026-01-01T00:00:00Z" }].to_json
      end
    end
    @server.mount_proc "/api/wallets/w1/balance" do |req, res|
      res.content_type = "application/json"
      res.body = [{ chain_id: "eip155:8453", token_address: "0x0", balance: "1000", decimals: 6 }].to_json
    end
    @server.mount_proc "/api/wallets/w1/fund" do |req, res|
      res.content_type = "application/json"
      res.body = { txHash: "0xabc" }.to_json
    end
    @server.mount_proc "/api/swap/quote" do |req, res|
      res.content_type = "application/json"
      res.body = { from_token: "0xA", to_token: "0xB", from_amount: "1000", to_amount: "2000", chain: "base" }.to_json
    end
    @server.mount_proc "/api/ai/generate" do |req, res|
      res.content_type = "application/json"
      res.body = { text: "hello", cost: "0.001" }.to_json
    end
    @server.mount_proc "/api/wallets" do |req, res|
      res.content_type = "application/json"
      if req.request_method == "POST"
        body = JSON.parse(req.body)
        if body["name"].nil? || body["name"].empty?
          res.status = 400
          res.body = { error: "bad request" }.to_json
        else
          res.body = { id: "w1", name: "demo", status: "active", created_at: "2026-01-01T00:00:00Z" }.to_json
        end
      else
        res.body = [{ id: "w1", name: "demo", status: "active", created_at: "2026-01-01T00:00:00Z" }].to_json
      end
    end
    Thread.new { @server.start }
    @port = @server.config[:Port]
    @client = Gradience::Client.new("http://127.0.0.1:#{@port}", api_token: "token")
  end

  def teardown
    @server.shutdown
  end

  def test_create_wallet
    wallet = @client.create_wallet("demo")
    assert_equal "w1", wallet["id"]
    assert_equal "demo", wallet["name"]
  end

  def test_list_wallets
    wallets = @client.list_wallets
    assert_equal 1, wallets.length
    assert_equal "w1", wallets[0]["id"]
  end

  def test_get_balance
    balances = @client.get_balance("w1")
    assert_equal "1000", balances[0]["balance"]
  end

  def test_fund_wallet
    res = @client.fund_wallet("w1", to: "0xto", amount: "1", chain: "base")
    assert_equal "0xabc", res["txHash"]
  end

  def test_swap_quote
    res = @client.swap_quote("w1", from_token: "0xA", to_token: "0xB", amount: "1000", chain: "base")
    assert_equal "2000", res["to_amount"]
  end

  def test_ai_generate
    res = @client.ai_generate("w1", model: "claude", prompt: "hi")
    assert_equal "hello", res["text"]
  end

  def test_error_response
    assert_raises(Gradience::Error) do
      @client.create_wallet("")
    end
  end
end

File: sdk/python/tests/test_client.py
import json
import unittest
from unittest.mock import patch, MagicMock
from gradience_sdk import GradienceClient, GradienceError


class FakeResponse:
    def __init__(self, status_code, json_data=None, text=None):
        self.status_code = status_code
        self._json = json_data
        self._text = text or json.dumps(json_data)
        self.ok = status_code < 400

    def json(self):
        return self._json

    def text(self):
        return self._text


class TestGradienceClient(unittest.TestCase):
    def setUp(self):
        self.client = GradienceClient("http://localhost:8080", api_token="token")

    @patch("gradience_sdk.client.requests.request")
    def test_create_wallet(self, mock_request):
        mock_request.return_value = FakeResponse(201, {"id": "w1", "name": "demo", "status": "active", "created_at": "2026-01-01T00:00:00Z"})
        wallet = self.client.create_wallet("demo")
        self.assertEqual(wallet["id"], "w1")
        mock_request.assert_called_once()
        args = mock_request.call_args
        self.assertEqual(args[0][0], "POST")
        self.assertIn("/api/wallets", args[0][1])

    @patch("gradience_sdk.client.requests.request")
    def test_list_wallets(self, mock_request):
        mock_request.return_value = FakeResponse(200, [{"id": "w1", "name": "demo", "status": "active", "created_at": "2026-01-01T00:00:00Z"}])
        wallets = self.client.list_wallets()
        self.assertEqual(len(wallets), 1)
        self.assertEqual(wallets[0]["id"], "w1")

    @patch("gradience_sdk.client.requests.request")
    def test_get_balance(self, mock_request):
        mock_request.return_value = FakeResponse(200, [{"chain_id": "eip155:8453", "token_address": "0x0", "balance": "1000", "decimals": 6}])
        balances = self.client.get_balance("w1")
        self.assertEqual(balances[0]["balance"], "1000")

    @patch("gradience_sdk.client.requests.request")
    def test_fund_wallet(self, mock_request):
        mock_request.return_value = FakeResponse(200, {"txHash": "0xabc"})
        res = self.client.fund_wallet("w1", "0xto", "1", "base")
        self.assertEqual(res["txHash"], "0xabc")

    @patch("gradience_sdk.client.requests.request")
    def test_swap_quote(self, mock_request):
        mock_request.return_value = FakeResponse(200, {"from_token": "0xA", "to_token": "0xB", "from_amount": "1000", "to_amount": "2000", "chain": "base"})
        quote = self.client.swap_quote("w1", {"from_token": "0xA", "to_token": "0xB", "amount": "1000", "chain": "base"})
        self.assertEqual(quote["to_amount"], "2000")

    @patch("gradience_sdk.client.requests.request")
    def test_ai_generate(self, mock_request):
        mock_request.return_value = FakeResponse(200, {"text": "hello", "cost": "0.001"})
        res = self.client.ai_generate("w1", "claude", "hi")
        self.assertEqual(res["text"], "hello")

    @patch("gradience_sdk.client.requests.request")
    def test_error_raises_gradience_error(self, mock_request):
        mock_request.return_value = FakeResponse(400, {"error": "bad request"})
        with self.assertRaises(GradienceError) as ctx:
            self.client.create_wallet("")
        self.assertEqual(ctx.exception.status_code, 400)
        self.assertEqual(str(ctx.exception), "bad request")

    @patch("gradience_sdk.client.requests.request")
    def test_create_policy(self, mock_request):
        mock_request.return_value = FakeResponse(200, {"policy_id": "p1"})
        res = self.client.create_policy("w1", "{}")
        self.assertEqual(res["policy_id"], "p1")

    @patch("gradience_sdk.client.requests.request")
    def test_export_audit(self, mock_request):
        mock_request.return_value = FakeResponse(200, [{"id": "a1", "action": "sign"}])
        logs = self.client.export_audit("w1", "json")
        self.assertIsInstance(logs, list)


if __name__ == "__main__":
    unittest.main()
